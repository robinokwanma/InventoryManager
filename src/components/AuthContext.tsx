import { createContext, useContext, useState } from "react";

interface AuthContextValue {
  authenticated: boolean;
  setAuthenticated: React.Dispatch<React.SetStateAction<boolean>>;
}

const AuthContext = createContext<AuthContextValue | undefined>(undefined);

export const AuthProvider: React.FC = ({ children }) => {
  const [authenticated, setAuthenticated] = useState<boolean>(false);

  return (
    <AuthContext.Provider value={{ authenticated, setAuthenticated }}>
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
};
