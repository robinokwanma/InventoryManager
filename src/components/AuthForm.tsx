import React, { useState, useEffect, useRef } from "react";
import AsyncStorage from "@react-native-async-storage/async-storage";
import {
  View,
  Text,
  TextInput,
  Button,
  StyleSheet,
  FlatList,
  TouchableOpacity,
} from "react-native";

interface AuthFormProps {
  loginForm: LoginFormState;
  setLoginForm: (state: LoginFormState) => void;
  handleLogin: () => void;
}

export default function AuthForm({
  loginForm,
  setLoginForm,
  handleLogin,
}: AuthFormProps) {
  const renderLoginForm = () => (
    <View style={styles.loginContainer}>
      <Text style={styles.loginTitle}>Login</Text>
      <TextInput
        style={styles.input}
        placeholder="Email"
        onChangeText={(text) => setLoginForm({ ...loginForm, email: text })}
        value={loginForm.email}
        keyboardType="email-address"
        autoCapitalize="none"
      />
      <TextInput
        style={styles.input}
        placeholder="Password"
        onChangeText={(text) => setLoginForm({ ...loginForm, password: text })}
        value={loginForm.password}
        secureTextEntry={true}
      />
      <Button title="Login" onPress={handleLogin} />
    </View>
  );

  return renderLoginForm();
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
    marginTop: 80,
  },
  input: {
    borderColor: "gray",
    borderWidth: 1,
    marginBottom: 10,
    padding: 5,
  },
  item: {
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "center",
    marginBottom: 10,
    padding: 10,
    borderColor: "gray",
    borderWidth: 1,
  },
  loginContainer: {
    flex: 1,
    padding: 20,
    justifyContent: "center",
  },
  loginTitle: {
    fontSize: 24,
    marginBottom: 20,
    alignSelf: "center",
  },
});
