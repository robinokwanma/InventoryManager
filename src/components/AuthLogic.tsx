import AsyncStorage from "@react-native-async-storage/async-storage";

export const login = async (email: string, password: string) => {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

  if (emailRegex.test(email) && password.length > 0) {
    await AsyncStorage.setItem("user", JSON.stringify({ email, password }));
    return true;
  } else {
    throw new Error(
      "Error during login, try again. pleasue use a valid email and make sure your password has at least 4 characters"
    );
  }
};

export const logout = async () => {
  await AsyncStorage.removeItem("user");
};

export const isLoggedIn = async (): Promise<boolean> => {
  const user = await AsyncStorage.getItem("user");
  return user !== null;
};
