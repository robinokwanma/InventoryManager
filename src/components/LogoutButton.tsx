import React from "react";
import { Alert, Image, Text, TouchableOpacity } from "react-native";
import { useAuth } from "./AuthContext";
import AsyncStorage from "@react-native-async-storage/async-storage";

const LogoutButton = () => {
  const { setAuthenticated } = useAuth();

  const handleLogout = () => {
    Alert.alert(
      "Logout",
      "Are you sure you want to log out?",
      [
        {
          text: "Cancel",
          style: "cancel",
        },
        {
          text: "Logout",
          onPress: async () => {
            // Call logout logic
            setAuthenticated(false);
            try {
              await AsyncStorage.removeItem("authenticated");
            } catch (error) {
              console.log(error);
            }
          },
        },
      ],
      { cancelable: true }
    );
  };

  return (
    <TouchableOpacity style={{ marginRight: 30 }} onPress={handleLogout}>
      <Image
        source={require("../../assets/logout.png")}
        resizeMode="contain"
        style={{
          width: 25,
          height: 25,
          marginRight: 5,
          tintColor: "red",
    
        }}
      />
      <Text style={{ fontSize: 18, fontWeight: "bold", marginBottom: 15 }}>Logout</Text>
    </TouchableOpacity>
    // <TouchableOpacity style={{ flexDirection: "row", alignItems: "center" }} onPress={handleLogout}>
    //   <Image source={require("../../assets/logout.png")} style={{ width: 20, height: 20, marginRight: 5 }} />
    //   <Text>Logout</Text>
    // </TouchableOpacity>
  );
};

export default LogoutButton;
