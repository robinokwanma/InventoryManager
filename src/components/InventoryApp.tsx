import ItemsList from "../screens/ItemsList";
import AddItems from "../screens/AddItems";
import React, { useState, useEffect } from "react";
import { View } from "react-native";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { isLoggedIn } from "../components/AuthLogic";
import { useAuth } from "../components/AuthContext";

interface Item {
  id: number;
  name: string;
  price: number;
  totalStock: number;
  description: string;
}

const InventoryApp = () => {
  const [items, setItems] = useState<Item[]>([]);
  const { authenticated, setAuthenticated } = useAuth();

  useEffect(() => {
    const loadItems = async () => {
      try {
        const storedItems = await AsyncStorage.getItem("inventory");
        if (storedItems) {
          setItems(JSON.parse(storedItems));
        }
      } catch (error) {
        console.log(error);
      }

      const loggedIn = await isLoggedIn();
      setAuthenticated(loggedIn);
    };

    loadItems();
  }, []);

  const handleAddItem = (newItem: Item) => {
    setItems((prevItems) => [...prevItems, newItem]);
  };

  return (
    <View>
      <AddItems
        items={items}
        setItems={setItems}
        authenticated={authenticated}
        setAuthenticated={setAuthenticated}
        onAddItem={handleAddItem}
      />
      <ItemsList
        items={items}
        setItems={setItems}
        authenticated={authenticated}
        setAuthenticated={setAuthenticated}
      />
    </View>
  );
};

export default InventoryApp;
