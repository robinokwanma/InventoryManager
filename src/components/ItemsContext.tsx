import React, { createContext, useState, useContext } from "react";
import { Item } from "../screens/ItemsList"; // Import the Item interface

interface ItemsContextProps {
  items: Item[];
  setItems: React.Dispatch<React.SetStateAction<Item[]>>;
}

const ItemsContext = createContext<ItemsContextProps>({
  items: [],
  setItems: () => {},
});

export const useItems = () => useContext(ItemsContext);

export const ItemsProvider: React.FC = ({ children }) => {
  const [items, setItems] = useState<Item[]>([]);

  return (
    <ItemsContext.Provider value={{ items, setItems }}>
      {children}
    </ItemsContext.Provider>
  );
};
