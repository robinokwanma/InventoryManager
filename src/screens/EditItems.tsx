import React, { useState, useEffect, useRef } from "react";
import {
  View,
  Text,
  TextInput,
  Button,
  StyleSheet,
  FlatList,
  Image,
  Alert,
} from "react-native";
import { useFocusEffect } from "@react-navigation/native";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { useAuth } from "../components/AuthContext";
import { useNavigation } from "@react-navigation/native";
import { login, logout, isLoggedIn } from "../components/AuthLogic";
import AuthForm from "../components/AuthForm";
import { Item } from "./ItemsList";
import LogoutButton from "../components/LogoutButton";
import { useItems } from "../components/ItemsContext";

interface EditItemsProps {
  isTest?: boolean;
  items: Item[];
  setItems: React.Dispatch<React.SetStateAction<Item[]>>;
  authenticated: boolean;
  setAuthenticated: React.Dispatch<React.SetStateAction<boolean>>;
}

const EditItems = () => {
  const { items, setItems } = useItems();
  const { authenticated, setAuthenticated } = useAuth();
  const navigation = useNavigation();
  const [searchQuery, setSearchQuery] = useState("");
  const [filteredItems, setFilteredItems] = useState(items);
  const [loginForm, setLoginForm] = useState<LoginFormState>({
    email: "",
    password: "",
  });

  useEffect(() => {
    setFilteredItems(
      items.filter((item) =>
        item.name.toLowerCase().includes(searchQuery.toLowerCase())
      )
    );
  }, [searchQuery, items]);

  useFocusEffect(
    React.useCallback(() => {
      const loadItems = async () => {
        const storedItems = await AsyncStorage.getItem("inventory");
        try {
          if (storedItems) {
            setItems(JSON.parse(storedItems));
          }
        } catch (error) {
          console.log(error);
        }
      };

      loadItems();

      // Returning a cleanup function to avoid memory leaks when the component is unmounted
      return () => {};
    }, [])
  );

  const editItem = (id: number) => {
    navigation.navigate("EditItem", { itemId: id, items: items });
  };

  const handleLogin = async () => {
    if (loginForm.email && loginForm.password) {
      try {
        await login(loginForm.email, loginForm.password);
        setAuthenticated(true);
      } catch (error) {
        alert(error.message);
      }
    } else {
      alert("Please enter email and password");
    }
  };

  const deleteItem = async (id: number) => {
    Alert.alert("Delete Item", "Are you sure you want to delete this item?", [
      {
        text: "Cancel",
        style: "cancel",
      },
      {
        text: "Delete",
        onPress: async () => {
          const filteredItems = items.filter((item) => item.id !== id);
          setItems(filteredItems);

          try {
            await AsyncStorage.setItem(
              "inventory",
              JSON.stringify(filteredItems)
            );
          } catch (error) {
            console.log(error);
          }
        },
        style: "destructive",
      },
    ]);
  };

  const renderItem = ({ item }: { item: Item }) => (
    <View style={styles.item} >
      <Text style={styles.title}>{item.name}</Text>
      <Image
        style={styles.image}
        source={{ uri: "https://via.placeholder.com/150" }}
      />
      <Text>Price: ${item.id}</Text>
      <Text>{item.description}</Text>
      <View style={styles.priceStockContainer}>
        <Text>Total Stock: {item.totalStock}</Text>
        <Text>Price: ${item.price}</Text>
        
      </View>
      <View style={styles.buttonContainer} >
        <Button title="Edit" onPress={() => editItem(item.id)} />
        <Button title="Delete" onPress={() => deleteItem(item.id)} testID="delete-button"/>
      </View>
    </View>
  );

  React.useLayoutEffect(() => {
    navigation.setOptions({
      headerRight: () => <LogoutButton />,
    });
  }, [navigation]);

  return (
    <View style={styles.container}>
      {authenticated ? (
        <>
          <Text style={{ marginBottom: 5 }}>
            <Text>Search </Text>

            <Image
              source={require("../../assets/search.png")}
              resizeMode="contain"
              style={{
                width: 25,
                height: 25,
                marginRight: 10,
              }}
            />
          </Text>
          <TextInput
            style={styles.searchBar}
            placeholder="Search n Edit..."
            onChangeText={setSearchQuery}
            value={searchQuery}
          />
          {filteredItems.length > 0 ? (
            <FlatList
              data={filteredItems}
              keyExtractor={(item) => (item.id ?? 0).toString()}
              renderItem={renderItem}
            />
          ) : (
            <Text style={styles.noMatches}>No matches</Text>
          )}
        </>
      ) : (
        <AuthForm
          loginForm={loginForm}
          setLoginForm={setLoginForm}
          handleLogin={handleLogin}
        />
      )}
    </View>
  );
};

export default EditItems;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
  },
  item: {
    marginBottom: 20,
    padding: 10,
    borderColor: "gray",
    borderWidth: 1,
  },
  title: {
    fontSize: 24,
    fontWeight: "bold",
  },
  image: {
    width: "100%",
    height: 150,
    resizeMode: "cover",
    marginTop: 10,
  },
  priceStockContainer: {
    flexDirection: "row",
    justifyContent: "space-between",
    marginTop: 10,
  },
  buttonContainer: {
    flexDirection: "row",
    justifyContent: "space-between",
    marginTop: 10,
  },
  searchBar: {
    borderColor: "gray",
    borderWidth: 1,
    marginBottom: 20,
    padding: 5,
    flexDirection: "row",
    alignItems: "center",
    borderRadius: 10,
    paddingHorizontal: 10,
  },
  noMatches: {
    fontSize: 18,
    textAlign: "center",
    marginTop: 20,
  },
});
