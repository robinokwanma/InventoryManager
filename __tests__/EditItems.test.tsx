import React from "react";
import { render , fireEvent, waitFor} from "@testing-library/react-native";
import { describe, expect, it } from "@jest/globals";
import { NavigationContainer } from "@react-navigation/native";
import { AuthProvider } from '../src/components/AuthContext';
import { useNavigation } from "@react-navigation/native";
import EditItems from "../src/screens/EditItems";
import { Alert } from "react-native";
import TestNavigator from "./TestNavigator";

// Snap Shot Test STARTS Here

//Mock  Navigation Module
jest.mock("@react-navigation/native", () => ({
  ...jest.requireActual("@react-navigation/native"),
  useNavigation: jest.fn(),
}));

describe("EditItems screen", () => {

  beforeEach(() => {
    // Provide a dummy implementation of setOptions for the test
    (useNavigation as jest.Mock).mockReturnValue({
      setOptions: jest.fn(),
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly", () => {
    const { toJSON } = render(
      <NavigationContainer>
        <AuthProvider>
          <EditItems />
        </AuthProvider>
      </NavigationContainer>
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
// Snap Shot Test ENDS Here



// Confirmation Pop-up on Delete Test STARTS Here

// Mock the Alert module
jest.mock("react-native/Libraries/Alert/Alert", () => ({
  alert: jest.fn(),
}));
jest.mock("../src/components/ItemsContext");

it("shows confirmation pop-up when trying to delete an item", async () => {
  const { queryByTestId, findByTestId } = render(
    <NavigationContainer>
      <AuthProvider>
      <TestNavigator
          screenOptions={{ headerShown: true }}
          Component={EditItems}
          props={{
            items: [
              {
                id: 1,
                name: "Item 1",
                description: "Item 1 description totk",
                totalStock: 10,
                price: 100,
              },
              {
                id: 2,
                name: "Item 2",
                description: "Item 2  svssrdf erbvd",
                totalStock: 5,
                price: 50,
              },
            ],
            setItems: () => {},
          }}
        >
        </TestNavigator>
      </AuthProvider>
    </NavigationContainer>
  );

  // Assuming your delete button has testID="deleteButton"
  const deleteButton =  queryByTestId("delete-button");
  fireEvent.press(deleteButton);

  // Check if the alert is called
  expect(Alert.alert).toHaveBeenCalled();

  // Check the title and message of the alert prompt
  const [alertTitle, alertMessage, actions] = (Alert.alert as jest.Mock).mock.calls[0];

  expect(alertTitle).toBe("Confirm Deletion");
  expect(alertMessage).toBe("Are you sure you want to delete this item?");
  expect(actions.length).toBe(2);
  expect(actions[0].text).toBe("Cancel");
  expect(actions[1].text).toBe("Delete");
});
