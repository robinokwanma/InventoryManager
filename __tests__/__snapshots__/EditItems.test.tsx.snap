// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditItems screen renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
      "padding": 20,
    }
  }
>
  <View
    style={
      {
        "flex": 1,
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <Text
      style={
        {
          "alignSelf": "center",
          "fontSize": 24,
          "marginBottom": 20,
        }
      }
    >
      Login
    </Text>
    <TextInput
      autoCapitalize="none"
      keyboardType="email-address"
      onChangeText={[Function]}
      placeholder="Email"
      style={
        {
          "borderColor": "gray",
          "borderWidth": 1,
          "marginBottom": 10,
          "padding": 5,
        }
      }
      value=""
    />
    <TextInput
      onChangeText={[Function]}
      placeholder="Password"
      secureTextEntry={true}
      style={
        {
          "borderColor": "gray",
          "borderWidth": 1,
          "marginBottom": 10,
          "padding": 5,
        }
      }
      value=""
    />
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Login
        </Text>
      </View>
    </View>
  </View>
</View>
`;
